### ISSUE TITLE
Encrypted disks with throw-away keys for /tmp

### ISSUE THREAD

**Issue by letmaik** [lsds/sgx-lkl#238]
Currently, AFAIK, /tmp is a tmpfs that is counted towards enclave memory. It would be nice to have the option of explicitly adding an encrypted /tmp disk of large size that's dynamically generated with a throw-away key that wouldn't be stored anywhere.

**Comment by davidchisnall** [comment]
It's slightly worse than that.  The fact that we are currently not using LKL's mmap means that tmpfs is coming from the (very scarce)  32MiB area allocated to LKL.  #187 would improve this somewhat, though it would be nice to be able to use the increased space as buffer cache and swap to an encrypted (+ dm-integrity) backing store.

**Comment by letmaik** [comment]
Right, we could use an ad-hoc encrypted /tmp disk also as swap by storing a swap file on it and configuring it via the app config, which is probably the easiest. If you want to use an actual swap partition then this would probably require more changes and would likely be another encrypted disk. I don't know about the tradeoffs between the two approaches (file vs partition).

**Comment by letmaik** [comment]
There are other temp folders as well. I found http://0pointer.net/blog/projects/tmp.html and https://specifications.freedesktop.org/basedir-spec/latest/ar01s03.html a useful read.

I could imagine that some apps use `$XDG_CACHE_HOME` (=`$HOME/.cache` if env var missing) or `$XDG_DATA_HOME` (=`$HOME/.local/share` if env var missing) for larger files.

`$XDG_RUNTIME_DIR` (on my system `/run/user/<uid>`) is typically tmpfs and used for very small files like communication primitives. The spec says that if `$XDG_RUNTIME_DIR` is missing then the app should print a warning and use a replacement with similar capabilities.

`/tmp` seems to be deprecated/legacy by now, but obviously still heavily used, and Python's `tempfile` module uses it as first fall-back if `$TMPDIR`/`$TEMP`/`$TMP` are not defined (which they are typically not). Because of this I don't think we can assume that `/tmp` will in practice only be used for small files.

With Docker images, setting environment variables and using the right storage would be done with the Dockerfile (directly or via a base image) and the `docker run` command (e.g. you can mount `/tmp` as a `tmpfs`).

Once throw-away encrypted disks are supported I think this is mostly a tooling and documentation matter, maybe showing some best practices.

**Comment by letmaik** [comment]
SGX-LKL currently mounts `/tmp` and `/run` automatically, which I think is a bit of an outlier. Docker only mounts `/proc`, `/dev`, `/dev/shm`, `/sys`. `/tmp` and `/run` are part of the base image. Since we don't have an overlayfs, `/tmp` and `/run` would be unwritable if the root disk is read-only and they are not mounted as `tmpfs`, so it does make some sense, since by default things should just work. I think though this should only be a default and be customizable.

For context, this is [how](https://docs.docker.com/engine/reference/run/#tmpfs-mount-tmpfs-filesystems) Docker allows to use a `tmpfs` instead of the overlay:
```
docker run -d --tmpfs /run:rw,noexec,nosuid,size=65536k my_image
```
I've seen people do the same for `/tmp`.

I propose the following:
- On start-up, (like now) the root disk and the special folders are mounted, (unlike now) excluding `/tmp` and `/run`.
- Then, all other disks are mounted (like now).
- Finally, a new function checks for each of `/tmp` and `/run` whether they were specified as disks in the app config and if not then they are mounted (like now) as `tmpfs`.
- In the long term, if at some point SGX-LKL should get support for an `overlayfs`, then I would remove the previous point such that `/tmp` and `/run` stay on the rootfs and can be written to via the `overlayfs`, like Docker is doing it.
- The disk config options should be extended such that it is possible to replicate the default `tmpfs` mounting of `/tmp` and `/run`. This allows to use `tmpfs` for other locations as well, but also allows to change the size used for `/tmp` and `/run` if they should stay as `tmpfs` instead of a physical disk.
- The disk config options should be extended to support throw-away encryption keys for non-tmpfs disks.

@prp @davidchisnall Does this make sense?

**Comment by prp** [comment]
@letmaik Currently, the mounts in SGX-LKL are as follows:
```
/dev/vda on / type ext4 (rw,relatime)
devtmpfs on /dev type devtmpfs (rw,relatime)
tmpfs on /dev/shm type tmpfs (rw,relatime)
tmpfs on /tmp type tmpfs (rw,relatime,mode=777)
tmpfs on /mnt type tmpfs (rw,relatime,mode=777)
none on /sys type sysfs (rw,relatime)
tmpfs on /run type tmpfs (rw,relatime,mode=700)
proc on /proc type proc (rw,relatime)
```

So `/tmp` and `/run` are mounted as `tmpfs`. Since this uses LKL memory, the storage space would be rather limited, so we definitely need a mechanism to override, as you suggest.

I'm wondering if it would make sense to generalise this and provide something that looks like `fstab` as part of the app_config, so any of the mounts can be overridden (as opposed to just `/tmp` and `/run`).

**Comment by letmaik** [comment]
Distros use systemd nowadays and the default mounts are actually hard-coded: https://github.com/systemd/systemd/blob/b4e1563ffb12c50ff89c063e721b76bfc89407b7/src/core/mount-setup.c#L63-L110. It also supports *.mount files which is an alternative to fstab apparently. fstab is also used at some point, and then there's initrd as well, which seems to have an effect on how systemd works if mounts exist already. It's all quite confusing, especially after reading https://bugs.launchpad.net/ubuntu/+source/initramfs-tools/+bug/1540803.

In general, I think we can take inspiration from fstab/systemd but I don't think there's a single right answer.
